pkg/query/optimize.go

```diff
@@ -0,0 +1,97 @@
+package query
+
+// TypedOptimizerFunc is a function that transforms an iterator of a specific type T
+// into a potentially optimized iterator. It returns the optimized iterator, a boolean
+// indicating whether any optimization was performed, and an error if the optimization failed.
+//
+// The type parameter T constrains the function to operate only on specific iterator types,
+// providing compile-time type safety when creating typed optimizers.
+type TypedOptimizerFunc[T Iterator] func(it T) (Iterator, bool, error)
+
+// OptimizerFunc is a type-erased wrapper around TypedOptimizerFunc[T] that can be
+// stored in a homogeneous list while maintaining type safety at runtime.
+type OptimizerFunc func(it Iterator) (Iterator, bool, error)
+
+// WrapOptimizer wraps a typed TypedOptimizerFunc[T] into a type-erased OptimizerFunc.
+// This allows optimizer functions for different concrete iterator types to be stored
+// together in a heterogeneous list.
+func WrapOptimizer[T Iterator](fn TypedOptimizerFunc[T]) OptimizerFunc {
+	return func(it Iterator) (Iterator, bool, error) {
+		if v, ok := it.(T); ok {
+			return fn(v)
+		}
+		return it, false, nil
+	}
+}
+
+// StaticOptimizations is a list of optimization functions that can be safely applied
+// to any iterator tree without needing runtime information or context.
+var StaticOptimizations = []OptimizerFunc{
+	RemoveNullIterators,
+	ElideSingletonUnionAndIntersection,
```

Comment:
Had to google what "elide" meant 😅  Now that I looked at the code I would have called this "collapse"